### PR TITLE
Embeddings

### PR DESCRIPTION
**Pull request by lmeyerov**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,7 @@ lib64/
 parts/
 sdist/
 var/
+data/
```
Comment (lmeyerov) [inline .gitignore:24]: let's not do git-lfs in a public repo as will likely be a big hurdle for novice contributors

File: ml/dgl_utils.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
+# classes for converting a dataframe or Graphistry Plottable into a DGL
+import pandas as pd
+import dgl
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import umap
+import numpy as np
```
Comment (lmeyerov) [inline ml/dgl_utils.py:8]: just a style thing, try to do one-line & sorted imports

File: ml/dgl_utils.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+FORMAT = "[%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
+logging.basicConfig(format=FORMAT)
+logger.setLevel(logging.DEBUG)
+
+
+# ndata_fields = ['feature', 'label', 'train_mask', 'test_mask', 'validation_mask']
+# edata_fields = ['feature', 'relationship', 'train_mask', 'test_mask', 'validation_mask']
+
+
+def get_vectorizer(name):
```
Comment (lmeyerov) [inline ml/dgl_utils.py:26]: we try to do py3 mypy-style type annotations ("`def get_vectorizer(name: str) -> ???`") , esp, for anything public (and full doc strings then)

File: ml/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+scikit-learn
+matplotlib
```
Comment (lmeyerov) [inline ml/requirements.txt:2]: Interesting, maybe we add a few things to https://github.com/graphistry/pygraphistry/blob/53bd7a779b9efc216301fcee94b493de9184cbc2/setup.py#L16

like `ml` (umap, ...) + `ai` (dgl)

File: ml/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+seaborn
```
Comment (lmeyerov) [inline ml/requirements.txt:3]: if possible, let's not include seaborn here, and instead, maybe our notebook dockerfile

File: ml/utils.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+
+
+# #################################################################################################
+#
+#   Finding subgraphs within a large graph.
+#
+# #################################################################################################
+
+
+def search_to_df(word, col, df):
```
Comment (lmeyerov) [inline ml/utils.py:127]: Vectorize impl, esp for GPU:

Ex: `df[df['month'].str.contains('Ju')]`

File: ml/utils.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    :param search_term:
+    :param src:
+    :param dst:
+    :param edf:
+    :return: pandas.DataFrame
+    """
+    logger.info(f"Finding {search_term} in both {src} and {dst} columns")
+    tdf = pd.concat(
+        [search_to_df(search_term, src, edf), search_to_df(search_term, dst, edf)],
```
Comment (lmeyerov) [inline ml/utils.py:158]: ... is this right?

it seems like some ~NLP helpers for pattern search are getting added (cypher-lite)... we may want to keep as internal pygraphistry api for now as likely unstable

File: ml/utils.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    word = word.lower()
+    try:
+        res = df[df[col].apply(lambda x: word in str(x).lower())]
+    except TypeError as e:
+        logger.error(e)
+        return df
+    return res
+
+
+def get_nearest(search_term, src, dst, edf):
```
Comment (lmeyerov) [inline ml/utils.py:148]: often passing src/dst/edf: better to pass `g`

src = g._source
dst = g._destination
edf = g._edges

File: ml/utils.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    gcols = pd.concat(
+        [tdf[dst], tdf[src]], axis=0
+    )  # get all node entries that show up in edge graph
+    ntdf = ndf[ndf[node_col].isin(gcols)]
+    g = graphistry.edges(tdf, src, dst).nodes(ntdf, node_col)
+    return g
+
+
+# lets make simple functions to find subgraphs by search_term = a given node
+def get_milieu_graph_from_search(search_term, src, dst, edf, both=False):
```
Comment (lmeyerov) [inline ml/utils.py:185]: is milieu a technical term? is it different from neighborhood?

File: ml/utils.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+    srclist = df[src]
+    dstlist = df[dst]
+    cnt = Counter(pd.concat([srclist, dstlist], axis=0))
+    ordered_nodes_dict = {k: i for i, (k, c) in enumerate(cnt.most_common())}
+    df[config.SRC] = df[src].apply(lambda x: ordered_nodes_dict[x])
+    df[config.DST] = df[dst].apply(lambda x: ordered_nodes_dict[x])
+    return df, ordered_nodes_dict
+
+
+def pandas2dgl(df, src, dst, weight_col=None, device="cpu"):
```
Comment (lmeyerov) [inline ml/utils.py:285]: interesting... curious from a cudf/cugraph side as well

File: ml/dgl_utils.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    if y_enc is not None:
+        y_enc = torch.tensor(y_enc.values)
+        data = {config.FEATURE: torch.tensor(X_enc.values), config.TARGET: y_enc}
+    else:
+        data = {config.FEATURE: torch.tensor(X_enc.values)}
+    encoders = {config.FEATURE: sup_vec, config.TARGET: sup_label}
+    return data, encoders
+
+
+def scatterplot(ux, color_labels=None):
```
Comment (lmeyerov) [inline ml/dgl_utils.py:80]: let's drop any non-graphistry viz from pygraphistry: people can take the df's and plot with something else if they don't want pygraphistry's

for the case of a scatterplot, can use a graphistry plot with node x/y positions and play=0, see demos folders

File: ml/dgl_utils.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        if not hasattr(self, "ndata"):
+            self._node_featurization()
+            # first add the ndata
+            self.graph.ndata.update(self.ndata)
+        if not hasattr(self, "edata"):
+            self._edge_featurization()
+            # then add any edata
+            self.graph.edata.update(self.edata)
+
+    def umap(self, scale=False, show=False, color_labels=None):
```
Comment (lmeyerov) [inline ml/dgl_utils.py:219]: this seems tied to DGL... but needn't be, right? `g.nodes(...).umap(engine='umap_learn').plot()`

File: ml/dgl_utils.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+                    logger.info(f"Zscaling matrix")
+                    X_node = (X_node - X_node.mean(0)) / X_node.std(0)
+                logger.info(f"Fitting Umap over matrix of size {X_node.shape}")
+                # res = self._umapper.fit_transform(X_node, y_node)
+                res = self._umapper.fit_transform(X_node)
+                self._embedding = res
+                if show:
+                    scatterplot(res, color_labels=color_labels)
+        else:
+            logger.info(f"There are no Node Level embeddings to fit Umap over")
```
Comment (lmeyerov) [inline ml/dgl_utils.py:237]: see reference umap_learn nb for getting edges + plotting with graphistry

File: ml/dgl_utils.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+
+        # TODO add Edge Embedding
+
```
Comment (lmeyerov) [inline ml/dgl_utils.py:240]: see reference umap_learn nb for getting edges + plotting with graphistry

File: ml/dgl_utils.py [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+            self.graph.edata.update(self.edata)
+
+    def umap(self, scale=False, show=False, color_labels=None):
+        if hasattr(self, "graph"):
+            # y_node = None
+            # if config.TARGET in self.graph.ndata:
+            #     y_node = np.array(self.graph.ndata[config.TARGET])
+            if config.FEATURE in self.graph.ndata:
+                # take it out of a torch Tensor by wrapping it as ndarray
+                X_node = np.array(self.graph.ndata[config.FEATURE])
```
Comment (lmeyerov) [inline ml/dgl_utils.py:226]: We probably don't want to assume torch... so if also just dirty_cat or even just nodes/edges df, does that mean many cases here? or dirty_cat + gnn publish to some expected field with an expected format? (Some sort of new bindings, like Features, Labels, ...?)

File: ml/dgl_utils.py (hunk 12 above)
Comment (lmeyerov) [reply ml/dgl_utils.py:219]: our umap signature will be interesting!

-- passthrough args to engine (esp umap_learn vs cuml.umap)
-- smart defaults
-- info for cleaning vs settings vs optional training vs plotting...

File: .gitignore (hunk 1 above)
Comment (silkspace) [reply .gitignore:24]: Agreed. I'll update, and zip the sample data and send to you.

File: ml/dgl_utils.py (hunk 2 above)
Comment (silkspace) [reply ml/dgl_utils.py:8]: 100%

File: ml/utils.py (hunk 6 above)
Comment (silkspace) [reply ml/utils.py:127]: Great, thanks

File: ml/utils.py (hunk 8 above)
Comment (silkspace) [reply ml/utils.py:148]: Yes and this is what is needed in the next, when we just feed it a g:Plottable and it fills in things nicely.

File: ml/utils.py (hunk 10 above)
Comment (silkspace) [reply ml/utils.py:285]: Can also use pd.factorize(.., sort=True) but I wanted to label it by abundance of term (so 0 is most common, 1 is second most common edge-node, etc)

File: ml/constants.py [hunk 16]
```diff
@@ -2,7 +2,7 @@
 # source and destination labels for consistent pipe lining
 SRC = "_src"
 DST = "_dst"
-WEIGHT = '_weight'
+WEIGHT = "_weight"
```
Comment (lmeyerov) [inline ml/constants.py:5]: edge weight, node weight, .. ?

File: .gitignore (hunk 1 above)
Comment (lmeyerov) [reply .gitignore:24]: we can also do a separate optional gh repo, hotlink, or some other service folks use (s3, ...)

File: ml/__init__.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    API_KEY = "etc"
+
+
+# Get sample data
+def get_host_dataframe(nrows=100000):
+    df = pd.read_csv("pygraphistry/data/host_events.csv", nrows=nrows, index_col=0)
+    return df
+
+
+def get_netflow_dataframe(nrows=100000):
```
Comment (lmeyerov) [inline ml/__init__.py:14]: guessing these will go away

File: ml/constants.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
+# ############################################################
+# source and destination labels for consistent pipe lining
+SRC = "_src"
+DST = "_dst"
+WEIGHT = "_weight"
+
+# #############################
+# consistent clf pipelining
+FEATURE = "feature"
```
Comment (lmeyerov) [inline ml/constants.py:9]: should this be an array of feature cols, `NODE_FEATURE_COLS = [ ...]`, `EDGE_FEATURE_COLS = [ ... ]` ? 

and as a local prop of `PlotterBase` so folks can functionally manipulate variants?

Not sure yet what they're doing

File: ml/dgl_utils.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
+# classes for converting a dataframe or Graphistry Plottable into a DGL
+import numpy as np
+import pandas as pd
+import torch
```
Comment (lmeyerov) [inline ml/dgl_utils.py:4]: Couldn't find where imported, but we can't assume `torch` for baseline installs

Comment (lmeyerov) [reply ml/dgl_utils.py:4]: Also, try to do single-line imports... :) I think in the linter, not sure

File: ml/dgl_utils.py [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+
+    :param target: str, series, or dataframe
+    :param df: aux dataframe to use if target is a string
+    :param name: useful for when target is a series, it will become the column name of the returned dataframe
+    :return: dataframe
+    """
+    if type(target) == str:
+        # get the target from the dataframe
+        logger.info(f"Returning {name}-target {target} as DataFrame")
+        return pd.DataFrame({target: df[target].values}, index=df.index)
```
Comment (lmeyerov) [inline ml/dgl_utils.py:47]: this assumes `pd`. What happens if say `cudf`?

Comment (lmeyerov) [reply ml/dgl_utils.py:47]: ... generic version: `df[[target]]`

File: ml/dgl_utils.py [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+    if name == config.SKLEARN:
+        # TODO
+        return
+    logging.warning(
+        f"Vectorizer name must be one of [{config.DIRTY_CAT}, {config.SKLEARN}, ..]"
+    )
+    return
+
+
+def get_dataframe_for_target(target, df, name):
```
Comment (lmeyerov) [inline ml/dgl_utils.py:28]: not sure what `target` is so unsure of useful generalness here. ok to be minimal, esp. initially...

File: ml/dgl_utils.py [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+    if remove:
+        logger.info(f"Removing {remove_cols} columns from DataFrame")
+        tf = df.drop(columns=remove_cols)
+        return tf, y
+    return df, y
+
+
+def featurize_to_torch(df, y, vectorizer, remove=True):
+    """
+        Vectorize pandas DataFrames of model and target features into Torch compatible tensors
```
Comment (lmeyerov) [inline ml/dgl_utils.py:87]: for `cudf` DFs, looks like a bit of a dance, either:

-- stay on GPU: via dlpack: https://github.com/rapidsai/cudf/issues/5547
-- round trip: cudf -> cpu arrow -> torch

but maybe we wait for cudf until after 1.0

File: ml/models.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+from ml.utils import process_dirty_dataframes, calculate_column_similarity
+
+# from xgboost import XGBClassifier, plot_importance
+
+
+def get_data(model="botnet", nrows=10000):
+    y = None
+    X = None
+    if model == "botnet":
+        x_feats = [
```
Comment (lmeyerov) [inline ml/models.py:19]: guessing this is all demo code

File: ml/models.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
+# import optuna
```
Comment (lmeyerov) [inline ml/models.py:1]: ... interesting! Great example of where we'd want a remote-mode option!

File: ml/umap_utils.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+            return None
+        if y.shape[1] <= 1:
+            return y
+        else:
+            logger.warning(
+                f"Ignoring target column of shape {y.shape} as it is not one dimensional"
+            )
+            return None
+
+    def fit(self, X, y=None):
```
Comment (lmeyerov) [inline ml/umap_utils.py:38]: user kwargs?

File: ml/umap_utils.py [hunk 26]
```diff
@@ -0,0 +1,64 @@
+# a base class for UMAP to run on cpu or gpu
+import pandas as pd
+import umap
```
Comment (lmeyerov) [inline ml/umap_utils.py:3]: also `cuml.UMAP` if `engine="rapids"`

File: ml/umap_utils.py [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+                f"Ignoring target column of shape {y.shape} as it is not one dimensional"
+            )
+            return None
+
+    def fit(self, X, y=None):
+        y = self._check_target_is_one_dimensional(y)
+        self.reducer = self._reducer.fit(X, y)
+        self._edge_influence()
+
+    def transform(self, X):
```
Comment (lmeyerov) [inline ml/umap_utils.py:43]: user kwargs?

File: ml/umap_utils.py [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+        # After it is fit with y, transform doesn't use y
+        return self.reducer.transform(X)
+
+    def fit_transform(self, X, y=None):
```
Comment (lmeyerov) [inline ml/umap_utils.py:47]: user kwargs?

File: ml/utils.py (hunk 9 above)
Comment (lmeyerov) [reply ml/utils.py:192]: :) `get_k_hop(k=2)` and exn if `k != 2` :)

Comment (lmeyerov) [reply ml/utils.py:192]: Also, cugraph/networkx seem to have already: https://docs.rapids.ai/api/cugraph/stable/api_docs/api/cugraph.community.egonet.ego_graph.html?highlight=neighbors#cugraph-community-egonet-ego-graph

File: ml/utils.py [hunk 29]
```diff
@@ -0,0 +1,400 @@
...
+    :returns
+        df, pandas DataFrame with new edges.
+        ordered_nodes_dict, dict ordered from most common src and dst nodes.
+    """
+    srclist = df[src]
+    dstlist = df[dst]
+    cnt = Counter(pd.concat([srclist, dstlist], axis=0))
+    ordered_nodes_dict = {k: i for i, (k, c) in enumerate(cnt.most_common())}
+    df[config.SRC] = df[src].apply(lambda x: ordered_nodes_dict[x])
+    df[config.DST] = df[dst].apply(lambda x: ordered_nodes_dict[x])
```
Comment (lmeyerov) [inline ml/utils.py:288]: vectorize via join?

Comment (lmeyerov) [reply ml/utils.py:288]: cuGraph does this internally and exposes some methods, but that part of the API is unstable, so probably stay away from
